src/stream/src/executor/backfill/utils.rs

```diff
@@ -799,6 +805,37 @@ pub(crate) async fn persist_state<S: StateStore, const IS_REPLICATED: bool>(
     Ok(())
 }
 
+pub type BackfillRateLimiter =
+    RateLimiter<NotKeyed, InMemoryState, MonotonicClock, NoOpMiddleware<Instant>>;
+
+/// Creates a data chunk builder for snapshot read.
+/// If the `rate_limit` is smaller than `chunk_size`, it will take precedence.
+/// This is so we can partition snapshot read into smaller chunks than chunk size.
+pub fn create_builder_and_limiter(
+    rate_limit: Option<usize>,
+    chunk_size: usize,
+    data_types: Vec<DataType>,
+) -> (DataChunkBuilder, Option<BackfillRateLimiter>) {
+    if let Some(rate_limit) = rate_limit {
+        if rate_limit < chunk_size && rate_limit > 0 {
+            (
+                DataChunkBuilder::new(data_types, rate_limit),
+                Some(create_limiter(rate_limit)).flatten(),
+            )
+        } else if rate_limit >= chunk_size {
+            (
+                DataChunkBuilder::new(data_types, chunk_size),
+                Some(create_limiter(chunk_size)).flatten(),
```

Comment:
Why if rate limit larger than chunk size, we use chunk_size as rate limit 🤔

Reply:
Oh I messed it up, let me fix